internal/argocd/revision.go

```diff
@@ -2,107 +2,224 @@ package argocd
-		// Without an Application, we can't determine the desired revision.
-		return "", nil
-	case app.Spec.Source.Chart != "":
-		// This source points to a Helm chart.
+	case s.Chart != "":
 		// NB: This has to go first, as the repository URL can also point to
 		//     a Helm repository.
+		sourceLogger.Debug("Application source is a Helm chart")
 
-		// If there is a source update that targets app.Spec.Source, it might
-		// have its own ideas about the desired revision.
-		var targetPromoMechanism any
-		for i := range update.SourceUpdates {
-			sourceUpdate := &update.SourceUpdates[i]
-			if sourceUpdate.RepoURL == app.Spec.Source.RepoURL && sourceUpdate.Chart == app.Spec.Source.Chart {
-				targetPromoMechanism = sourceUpdate
-				break
-			}
-		}
-		if targetPromoMechanism == nil {
-			targetPromoMechanism = update
-		}
-		desiredOrigin := freight.GetDesiredOrigin(stage, targetPromoMechanism)
 		chart, err := freight.FindChart(
 			ctx,
 			cl,
 			stage,
 			desiredOrigin,
 			frght,
-			app.Spec.Source.RepoURL,
-			app.Spec.Source.Chart,
+			s.RepoURL,
+			s.Chart,
 		)
 		if err != nil {
-			return "", fmt.Errorf("error chart from repo %q: %w", app.Spec.Source.RepoURL, err)
+			sourceLogger.Error(err, "Error finding chart from repo")
+			return "", fmt.Errorf("error chart from repo %q: %w", s.RepoURL, err)
 		}
 		if chart == nil {
+			sourceLogger.Debug("Chart not found")
 			return "", nil
 		}
 		return chart.Version, nil
-	case app.Spec.Source.RepoURL != "":
-		// This source points to a Git repository.
-
-		// If there is a source update that targets app.Spec.Source, it might
-		// have its own ideas about the desired revision.
-		var targetPromoMechanism any
-		for i := range update.SourceUpdates {
-			sourceUpdate := &update.SourceUpdates[i]
-			if sourceUpdate.RepoURL == app.Spec.Source.RepoURL {
-				targetPromoMechanism = sourceUpdate
-				break
-			}
-		}
-		if targetPromoMechanism == nil {
-			targetPromoMechanism = update
-		}
-		desiredOrigin := freight.GetDesiredOrigin(stage, targetPromoMechanism)
+
+	case s.RepoURL != "":
+		sourceLogger.Debug("Application source is a Git repository")
 		commit, err := freight.FindCommit(
 			ctx,
 			cl,
 			stage,
 			desiredOrigin,
 			frght,
-			app.Spec.Source.RepoURL,
+			s.RepoURL,
 		)
 		if err != nil {
+			sourceLogger.Error(err, "Error finding commit from repo")
 			return "",
-				fmt.Errorf("error finding commit from repo %q: %w", app.Spec.Source.RepoURL, err)
+				fmt.Errorf("error finding commit from repo %q: %w", s.RepoURL, err)
 		}
 		if commit == nil {
+			sourceLogger.Debug("Commit not found")
 			return "", nil
 		}
 		if commit.HealthCheckCommit != "" {
 			return commit.HealthCheckCommit, nil
 		}
 		return commit.ID, nil
 	}
-	// If we end up here, no desired revision was found.
+
+	sourceLogger.Debug("Could not determine desired revision for application from this source.")
 	return "", nil
 }
+
+func getDesiredOriginForSource(
```

Comment:
I get what this function is doing, but its _name_ initially confused me since "desired origin" isn't really a function of an ApplicationSource. It's _purely_ a function of an update. Careful reading of the code reveals that you're using the source to help you find which target update to scrutinize further -- which is correct -- but if the function were, for instance, called `getTargetUpdate()` (and returned the target update), then the name of the function would much more closely match what it does and that would eliminate a lot of confusion.